src/layout_engine.rs

```diff
@@ -0,0 +1,206 @@
+		LayoutAbstractNode::Tag(tag) => compute_linear_layout(root, tag),
+		LayoutAbstractNode::Text(_) => {
+			// For text, we currently just use the hard-coded layout values
+			println!("Text layout: {:?}", root.borrow().layout_attributes());
+		},
+	}
+}
+
+fn compute_linear_layout(node: &LayoutNode, tag: &LayoutAbstractTag) {
+	// Extract the layout-specific attributes of this node
+	let mut attribs = node.borrow().layout_attributes();
+
+	// For the sake of example, we'll assume the main window has a fixed size
+	if tag.namespace == "window" && tag.name == "main" {
+		attribs.width = Dimension::AbsolutePx(1920.0);
+		attribs.height = Dimension::AbsolutePx(1080.0);
+	}
+
+	let main_axis = if tag.namespace == "" && tag.name == "row" { Axis::Horizontal } else { Axis::Vertical };
+	let cross_axis = main_axis.perpendicular();
+
+	// Go through all this node's children and partition them based on
+	// what sort of sizing they use along the main axis.
+	let mut fixed_size_children = Vec::new();
+	let mut percent_size_children = Vec::new();
+	let mut inner_size_children = Vec::new();
+	let mut remainder_size_children = Vec::new();
+	for child in node.children() {
+		let child_attribs = child.borrow().layout_attributes();
+		match child_attribs.size_along(main_axis) {
+			Dimension::AbsolutePx(size) => fixed_size_children.push((child, size)),
+			Dimension::Percent(percent) => percent_size_children.push((child, percent)),
+			Dimension::Inner => inner_size_children.push(child),
+			Dimension::PercentRemainder(percent) => remainder_size_children.push((child, percent)),
+			dimension => todo!("unsupported dimension type in layout: {:?}", dimension),
+		}
+		// TODO: on the cross axis, we'll need to expand to fit the children.
+		assert_eq!(child_attribs.size_along(cross_axis), Dimension::Percent(100.0), "only 100% on cross axis supported for now");
+	}
+
+	// The minimum size we need along the main axis to lay out all of the children
+	let mut min_main_axis_size = 0.0;
+
+	println!("Laying out fixed size children");
+	for (child, size) in fixed_size_children {
+		// TODO: recursively lay out children
+		//compute_layout(&child);
+		println!("Laid out child {:?} with {:?} size {}", child, main_axis, size);
+		min_main_axis_size += size;
+	}
+
+	// TODO: lay out percent size children
+
+	// TODO: lay out inner size children
+
+	// Check if we still fit in the parent container's defined size
+	if let Dimension::AbsolutePx(max_main_axis_size) = attribs.size_along(main_axis) {
+		if min_main_axis_size > max_main_axis_size {
+			todo!("main axis overflow is not supported yet")
+		}
+	}
+
+	// At the end, we lay out the elements which use the `@` specifier.
+	// They split up the remaining free space.
+	let free_space = {
+		if let Dimension::AbsolutePx(size) = attribs.size_along(main_axis) {
+			size - min_main_axis_size
+		}
+		else {
+			panic!("child layout uses free space percentage but parent doesn't have a well-defined size");
+		}
```

Comment:
This is a problematic scenario. What if the child wants to be 100@ of the free space of the parent, but the parent has an unbound 100% size?

The solution would be to always make sure the root of the hierarchy has a fixed size (for example, the window).

Reply:
The window will indeed be always called with absolute pixel values, like shown on line 1 of `window/main.xml`, equal to the dimensions of the window frame.